sql/src/main/java/io/crate/metadata/Signature.java

```diff
@@ -22,137 +22,310 @@
+         *
+         * @param predicate a predicate which represents the precondition
+         * @return a new operator with the predicate prepended
+         */
+        default SignatureOperator preCondition(Predicate<List<DataType>> predicate) {
+            return dataTypes -> predicate.test(dataTypes) ? this.apply(dataTypes) : null;
+        }
+
+        /**
+         * Adds a fallback to this operator. When this operator fails to match, the given fallback operator will
+         * be evaluated.
+         *
+         * @param fallback the fallback operator
+         * @return a new operator with the fallback appended
+         */
+        default SignatureOperator or(SignatureOperator fallback) {
+            return dataTypes -> {
+                List<DataType> n = this.apply(dataTypes);
+                if (n == null) {
+                    return fallback.apply(dataTypes);
+                }
+                return n;
+            };
         }
-        return builder.build();
+
+        /**
+         * Adds a secondary operator to this operator, which is executed with the result of this operator if a match
+         * occurs.
+         *
+         * @param secondary the operator which will be evaluated with this operator's success result
+         * @return a new operator with the secondary appended
+         */
+        default SignatureOperator and(SignatureOperator secondary) {
+            return dataTypes -> {
+                List<DataType> n = this.apply(dataTypes);
+                if (n != null) {
+                    return secondary.apply(n);
+                }
+                return null;
+            };
+        }
+
     }
 
-    private final ImmutableList<DataType> list;
+    public static final SignatureOperator EMPTY = dataTypes -> dataTypes.isEmpty() ? dataTypes : null;
+
+    public static final SignatureOperator SIGNATURES_SINGLE_NUMERIC = of(ArgMatcher.NUMERIC);
+    public static final SignatureOperator SIGNATURES_SINGLE_ANY = of(1);
+
+    public static final Predicate<DataType> IS_NULL = DataTypes.UNDEFINED::equals;
+
+
+    public interface ArgMatcher extends UnaryOperator<DataType> {
+
+        static ArgMatcher of(DataType... allowedTypes) {
+            return dt -> {
+                for (DataType allowedType : allowedTypes) {
+                    if (allowedType.equals(dt)) {
+                        return dt;
+                    }
+                }
+                return IS_NULL.test(dt) ? dt : null;
+            };
+        }
+
+        @SafeVarargs
+        static ArgMatcher of(final Predicate<DataType>... predicates) {
+            return dt -> {
+                for (Predicate<DataType> predicate : predicates) {
+                    if (predicate.test(dt)) {
+                        return dt;
+                    }
+                }
+                return IS_NULL.test(dt) ? dt : null;
+            };
+        }
+
+        static ArgMatcher rewriteTo(final DataType targetType) {
```

Comment:
Why are the other two methods called `of` and this one is `rewriteTo`, aren't they the same?